File: src/test/java/pl/mjedynak/idea/plugins/builder/factory/GoToBuilderPopupListFactoryTest.java
package pl.mjedynak.idea.plugins.builder.factory;

import static org.assertj.core.api.Assertions.assertThat;

import com.intellij.ui.ExpandedItemListCellRendererWrapper;
import com.intellij.ui.components.JBList;
import javax.swing.JList;
import org.junit.jupiter.api.Test;
import pl.mjedynak.idea.plugins.builder.renderer.ActionCellRenderer;

public class GoToBuilderPopupListFactoryTest {

    private final AbstractPopupListFactory popupListFactory = new GoToBuilderPopupListFactory();

    @Test
    void shouldCreateJBListWithActionCellRenderer() {
        // when
        JList<?> popupList = popupListFactory.getPopupList();

        // then
        assertThat(popupList).isInstanceOf(JBList.class);
        assertThat(popupList.getCellRenderer()).isInstanceOf(ExpandedItemListCellRendererWrapper.class);
        assertThat(((ExpandedItemListCellRendererWrapper<?>) popupList.getCellRenderer()).getWrappee())
                .isInstanceOf(ActionCellRenderer.class);
        assertThat(((JBList<?>) popupList).getItemsCount()).isEqualTo(1);
    }

    @Test
    void shouldLazilyInitializeCellRenderer() {
        // then
        assertThat(popupListFactory.getActionCellRenderer()).isNull();
    }

    @Test
    void shouldUseTheSameCellRendererForConsequentInvocations() {
        // when
        popupListFactory.getPopupList();
        ActionCellRenderer firstRenderer = popupListFactory.getActionCellRenderer();
        popupListFactory.getPopupList();
        ActionCellRenderer secondRenderer = popupListFactory.getActionCellRenderer();

        // then
        assertThat(firstRenderer).isSameAs(secondRenderer);
    }
}

File: src/test/kotlin/io/github/kingg22/kobra/builder/factory/GoToBuilderPopupListFactoryTest.kt
package io.github.kingg22.kobra.builder.factory

import com.intellij.ui.ExpandedItemListCellRendererWrapper
import com.intellij.ui.components.JBList
import io.github.kingg22.kobra.builder.renderer.ActionCellRenderer
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class GoToBuilderPopupListFactoryTest {
    private val popupListFactory: AbstractPopupListFactory = GoToBuilderPopupListFactory()

    @Test
    fun shouldCreateJBListWithActionCellRenderer() {
        // when
        val popupList = popupListFactory.popupList

        // then
        assertThat(popupList).isInstanceOf(JBList::class.java)
        assertThat(popupList.getCellRenderer()).isInstanceOf(ExpandedItemListCellRendererWrapper::class.java)
        assertThat(
            (popupList.getCellRenderer() as ExpandedItemListCellRendererWrapper<*>).wrappee,
        ).isInstanceOf(ActionCellRenderer::class.java)
        assertThat((popupList as JBList<*>).getItemsCount()).isEqualTo(1)
    }

    @Test
    fun shouldLazilyInitializeCellRenderer() {
        // then
        assertThat(popupListFactory.actionCellRenderer).isNull()
    }

    @Test
    fun shouldUseTheSameCellRendererForConsequentInvocations() {
        // when
        popupListFactory.popupList
        val firstRenderer = popupListFactory.actionCellRenderer
        popupListFactory.popupList
        val secondRenderer = popupListFactory.actionCellRenderer

        // then
        assertThat(firstRenderer).isSameAs(secondRenderer)
    }
}
